### DISCUSSION TITLE
Filer and seaweedfs-csi-driver security

### DISCUSSION THREAD

**Discussion by WhiteSash** [seaweedfs/seaweedfs#2132] [Q&A]
Hello
Is there any way to protect the filler's work when using the seaweedfs-csi-driver driver? At this moment, in the configuration of the seaweedfs-csi-driver, you need to specify the http port, so you cannot use the grpc so as not to open the filer service publicly. Also, is it possible to use certificates for authorization from the seaweedfs-csi-driver on the filler service to ensure security?

**Comment by chrislusf** [comment]
seaweedfs-csi-driver is using filer grpc port.

"is it possible to use certificates for authorization from the seaweedfs-csi-driver on the filler service"
Yes.

**Reply by WhiteSash** [reply]
You are my savior)) Сan you then clarify what settings need to be made?
I see 2 places in the seaweedfs-csi.yaml file where it is necessary to specify the filler's address and port. In both cases value is specified as an example: "192.168.2.7:8888"
Do I understand correctly that instead of the http port 8888, I need to explicitly specify port 18888 in the seaweedfs-csi.yaml file? Or is the http port 8888 indicated only for calculating the value of the port's grpc and all interaction is carried out through the grpc?
And where in the seaweedfs-csi.yaml configuration do I need to specify the client certificate for authorization in the filler service?
Extended answers would be greatly appreciated.

**Reply by chrislusf** [reply]
the actual port is using grpc with 10000+port.

use `weed scaffold -config=security` to generate the configurations. The certificates examples can be found here: https://github.com/chrislusf/seaweedfs/wiki/Security-Configuration

**Reply by WhiteSash** [reply]
I'm sorry, but I don't quite understand.
I applied the settings according to this page: https://github.com/chrislusf/seaweedfs/wiki/Security-Configuration
Configuration in security.toml is:
[grpc]
ca = "/etc/seaweedfs/pki/SeaweedFS_CA.crt"

[grpc.volume]
cert = "/etc/seaweedfs/pki/volume01.crt"
key  = "/etc/seaweedfs/pki/volume01.key"

[grpc.master]
cert = "/etc/seaweedfs/pki/master01.crt"
key  = "/etc/seaweedfs/pki/master01.key"

[grpc.filer]
cert = "/etc/seaweedfs/pki/filer01.crt"
key  = "/etc/seaweedfs/pki/filer01.key"

[grpc.client]
cert = "/etc/seaweedfs/pki/client01.crt"
key  = "/etc/seaweedfs/pki/client01.key"

After that I try to create sample PVC in the k8s cluster from https://github.com/seaweedfs/seaweedfs-csi-driver/blob/master/deploy/kubernetes/sample-seaweedfs-pvc.yaml, and I get the message: failed to provision volume with StorageClass "seaweedfs-storage": rpc error: code = Unknown desc = Error setting bucket metadata: mkdir / buckets / pvc-69ae89ef-6040-4fc6 -980a-568830199d45: CreateEntry: rpc error: code = Unavailable desc = connection closed
When I remove the security.toml configuration, the provisioning in k8s works out and we create PVC and PV successfully. 

In the configuration settings of the filler and master services, I use the -disableHttp parameter, but as far as I understand, this does not affect the work.
What is my mistake or what am I doing wrong?

**Reply by chrislusf** [reply]
These security configuration needs to be applied to all servers, not just mount.

**Reply by WhiteSash** [reply]
I don't really understand what this is about...
I have 3 servers, on each of them 3 application components work as different daemons - master, volume and filer. All servers have an identical security.toml file. There are certificates on each server. I can attach configurations if required. 
Do I need to change the yaml configuration of the csi provider? Add security.toml configuration somewhere in the seaweedfs pods?

**Reply by WhiteSash** [reply]
@chrislusf do you have any new information on this?

**Reply by chrislusf** [reply]
The client side, which is the csi driver, also needs the security.toml file.

**Reply by WhiteSash** [reply]
In my scenario, the seaweedfs services are deployed on the nodes of the Kubernetes cluster, the security.toml file is present on all nodes. And here the question is, what else needs to be configured and where?

**Reply by chrislusf** [reply]
Added a fix https://github.com/seaweedfs/seaweedfs-csi-driver/commit/5ac713b8b01a0e0a591a67d90a0bd13c48c968b7

The driver did not try to load the security.toml file before.

**Reply by WhiteSash** [reply]
Thank you for your prompt work. I will try in the coming days and sign off on the result ...

**Reply by WhiteSash** [reply]
@chrislusf as far as I understand, after editing the code, you need to rebuild and update the container images in the chrislusf/seaweedfs-csi-driver:latest ??

**Reply by aoberest** [reply]
@chrislusf Please rebuild and update the container images in the chrislusf/seaweedfs-csi-driver.
I have same issue.

**Reply by chrislusf** [reply]
updated the latest container image

**Reply by aoberest** [reply]
Thank you @chrislusf 


But nothing changed for me.

I recived same error
```
I0629 07:07:39     1 main.go:36] connect to filer 10.193.86.149:8888
I0629 07:07:39     1 driver.go:47] Driver: seaweedfs-csi-driver version: 1.0.0
I0629 07:07:39     1 driver.go:95] Enabling volume access mode: SINGLE_NODE_WRITER
I0629 07:07:39     1 driver.go:95] Enabling volume access mode: MULTI_NODE_MULTI_WRITER
I0629 07:07:39     1 driver.go:106] Enabling controller service capability: CREATE_DELETE_VOLUME
I0629 07:07:39     1 driver.go:106] Enabling controller service capability: PUBLISH_UNPUBLISH_VOLUME
I0629 07:07:39     1 server.go:92] Listening for connections on address: &net.UnixAddr{Name:"//var/lib/csi/sockets/pluginproxy/csi.sock", Net:"unix"}
I0629 07:07:40     1 filer_client.go:231] mkdir directory:"/buckets"  entry:{name:"pvc-37c13912-818d-4918-9fb1-e1c2087978c4"  is_directory:true  attributes:{mtime:1624950460  file_mode:2147484159  crtime:1624950460}}: CreateEntry: rpc error: code = Unavailable desc = connection closed
E0629 07:07:40     1 utils.go:56] GRPC error: Error setting bucket metadata: mkdir /buckets/pvc-37c13912-818d-4918-9fb1-e1c2087978c4: CreateEntry: rpc error: code = Unavailable desc = connection closed
I0629 07:07:41     1 filer_client.go:231] mkdir directory:"/buckets"  entry:{name:"pvc-37c13912-818d-4918-9fb1-e1c2087978c4"  is_directory:true  attributes:{mtime:1624950461  file_mode:2147484159  crtime:1624950461}}: CreateEntry: rpc error: code = Unavailable desc = connection closed
```
I wrote https://github.com/seaweedfs/seaweedfs-csi-driver/issues/25
